### ISSUE TITLE
[AI] 211122 오피스아워 이슈 : 모델 결과에 대한 이슈

### ISSUE THREAD

**Issue by JeongGod** [Whatcar/What-Car#46]
In GitLab by @JeongGyu on Nov 22, 2021, 17:47

## 개발 시 어려운 점을 아래 리스트 중에 체크해주세요.
| 이슈 Label| 해당 사항 체크(✅,❎) |
| ------ | ------ |
| 전체적인 아이디어, 프로젝트 진행 방향성에 관한 이슈 |  |
| To-do 기능 구현의 어려움 |  |
| 딥러닝 결과 해석의 어려움 |✅|
| 버그 |  |
| 서버 이슈 |  |
| 속도 이슈 |  |
| gitlab 이슈 |  |
| 그 외 |  |

- 이슈에 대한 자세한 설명 :

### 1. 개발 시 이슈가 무엇이었는지 예상 동작과 현재 동작에 대해 자세히 기입해주세요.
현재 detectron2 모델을 사용하여 진행중에 있습니다.  
지금 Acura라는 브랜드의 차종만 학습을 시켜서 잘 되는지 작동해보려고 했습니다.
하지만 현저하게 예측률이 떨어져서 무엇이 문제인지, 어디를 보강해야하는지 궁금합니다.

## Segmentation 후 box값 가져온 것 => 승욱님이 해결해주셨습니다.
![image](https://my-gitlab-bucket.s3.amazonaws.com/439202597/c8c3109b1cc27a7425c7b0f0d397e81e.png)

## 모델 구축 (mask_rcnn_R_50_FPN_3x)사용
```python
from detectron2.engine import DefaultTrainer

cfg = get_cfg()
cfg.merge_from_file(model_zoo.get_config_file("COCO-InstanceSegmentation/mask_rcnn_R_50_FPN_3x.yaml"))
cfg.DATASETS.TRAIN = ("acura_train",)
cfg.DATASETS.TEST = ()
cfg.DATALOADER.NUM_WORKERS = 2
cfg.MODEL.WEIGHTS = model_zoo.get_checkpoint_url("COCO-InstanceSegmentation/mask_rcnn_R_50_FPN_3x.yaml")
cfg.SOLVER.IMS_PER_BATCH = 2
cfg.SOLVER.BASE_LR = 0.00025
cfg.SOLVER.MAX_ITER = 1000
cfg.SOLVER.STEPS = []
cfg.MODEL.ROI_HEADS.BATCH_SIZE_PER_IMAGE = 256
cfg.MODEL.ROI_HEADS.NUM_CLASSES = 44

os.makedirs(cfg.OUTPUT_DIR, exist_ok=True)
trainer = DefaultTrainer(cfg) 
trainer.resume_or_load(resume=False)
trainer.train()
```
## 학습 막바지의 모습
```
[11/22 08:09:18 d2.utils.events]:  eta: 0:00:09  iter: 979  total_loss: 0.6102  loss_cls: 0.4497  loss_box_reg: 0.1035  loss_mask: 0.04493  loss_rpn_cls: 1.397e-09  loss_rpn_loc: 0.003315  time: 0.4821  data_time: 0.0028  lr: 0.00024476  max_mem: 2553M
[11/22 08:09:30 d2.utils.events]:  eta: 0:00:00  iter: 999  total_loss: 0.7619  loss_cls: 0.5444  loss_box_reg: 0.1698  loss_mask: 0.04809  loss_rpn_cls: 8.149e-10  loss_rpn_loc: 0.002915  time: 0.4825  data_time: 0.0028  lr: 0.00024975  max_mem: 2553M
```
개판입니다ㅠㅠ 이미지를 넣으면 예측률이 10%정도로 아예 제기능을 못하는 중입니다.

## 준비된 데이터셋
일단 테스트를 위해서 Acura라는 브랜드의 차만 학습을 시켜보려고 했습니다.  
데이터셋은 다음과 같습니다.  

![image](https://my-gitlab-bucket.s3.amazonaws.com/439202597/9731c11edab5485d6ee8dd186a3f6703.png)

Acura라는 브랜드에 대해서는 이러한 사진이 약 800개 정도 있습니다.  
이에대한 라벨은 `브랜드_모델명_연식`으로 나누어보니 라벨갯수만 43개정도가 됩니다.

![image](https://my-gitlab-bucket.s3.amazonaws.com/439202597/192d1b23de4be82dfe0da13d4b5bc219.png)

또, 이렇게 차의 뒷모습이나 차의 옆모습을 나타내는 사진도 존재합니다.

### 2. 이슈에 대한 원인을 기입해주세요.

1. 차의 앞,뒤,옆을 나타내는 사진이 모두 있다보니 이를 학습하는데 오차가 많이 생기는게 아닐까 생각이 듭니다.  
그래서 앞의 사진만 데이터를 처리해서 넣는것은 어떨까 하는 생각이 있습니다.
2. 갖고있는 데이터 개수에 비해서 라벨이 너무 많아 학습이 안되는 것 같습니다.  
데이터의 개수는 800개정도되는데 라벨이 40개정도 되다보니 하나의 라벨당 평균 20개정도의 데이터가 존재합니다.  
이 때문에 모델이 학습이 제대로 되지 않은 것 같습니다.

### 4. 이슈를 어떻게 고칠 것인지 자세히 설명해주세요. (해결 방법을 못 찾을 시, 5. 궁금한 점에 기입해주세요)

예시) DB에 접근하는 횟수가 적어질수록 계산식 속도가 빨라질 것이라 판단하여 계산식이 진행되는 동안에는 DB에 접근하지 않도록 수정할 예정입니다.

### 5. 그 외 궁금한 점을 기입해주세요.

1. 차의 앞 사진만 이용해서 학습을 시키면 조금 더 효율이 좋아질까요?
2. 데이터의 개수가 적은데 이를 증식시키는 방법은 agumentation이 있다고 들었습니다. 하지만 하나의 라벨당 평균 20장에 불과한 데이터를 증식시키는게 의미가 있을까 싶습니다.
3. 라벨을 줄이는 방향으로 간다면 `브랜드_모델명_연식` 에서 `브랜드_모델명`으로 바꾸려고 합니다. 그런데 연식연도를 뺀다면 연도별로 차종 외형이 많이 바뀌는때도 있는데 이는 학습에 지장이 없을지 궁금합니다.
4. 마지막으로 궁금한 점입니다! input img size가 현재 288*432로 고정입니다. 이로인해서 학습이 잘 안될 수도 있을까요? 이미지 resize를 해야할까요?

**Comment by JeongGod** [comment]
In GitLab by @JeongGyu on Nov 22, 2021, 17:56

changed the description

**Comment by JeongGod** [comment]
In GitLab by @JeongGyu on Nov 22, 2021, 18:01

changed title from **[AI] 211122 오피스아워 이슈 : {-input image size조정-}** to **[AI] 211122 오피스아워 이슈 : {+모델 결과에 대한 이슈+}**

**Comment by JeongGod** [comment]
In GitLab by @JeongGyu on Nov 22, 2021, 18:01

changed the description

**Comment by JeongGod** [comment]
In GitLab by @jschoi on Nov 23, 2021, 01:07

1. 차의 앞 사진만 이용해서 학습을 시키면 조금 더 효율이 좋아질까요?
효율은 좋아집니다. 그렇지만 서비스 관점에서 차량의 다른 모습을 보고도 차종을 알아 맞출 필요가 있을 때는 다각도의 사진을 학습시키는 것이 좋다고 생각합니다.

2. 데이터의 개수가 적은데 이를 증식시키는 방법은 agumentation이 있다고 들었습니다. 하지만 하나의 라벨당 평균 20장에 불과한 데이터를 증식시키는게 의미가 있을까 싶습니다.
증식 자체가 아주 효과적인! 요술방망이처럼 인식율을 끌어올려 주진 않습니다(인식율 40% 증가! 같이). 그렇지만 보통 10% 미만으로 어느 정도 효과를 인식율 증가 효과를 가져다 주긴 합니다. 그리고 말씀주신 것처럼 개수가 작을 때 하는 것이 augmentation이니 평균 20장 밖에 없을 수록 안하는 것보다는 하는 것이 저는 좋다고 생각이 듭니다. 

3. 라벨을 줄이는 방향으로 간다면 `브랜드_모델명_연식` 에서 `브랜드_모델명`으로 바꾸려고 합니다. 그런데 연식 연도를 뺀다면 연도별로 차종 외형이 많이 바뀌는때도 있는데 이는 학습에 지장이 없을지 궁금합니다. 
지장이 있습니다. 외형이 바뀐 다는 것은 이미지 특징이 다르다는 것인데 이를 같은 라벨군으로 학습시켜버리면 문제의 소지가 있습니다.
예를 들어 2020년도 K5와 2020년도 아반때가 똑같은 외형을 가지고 있다고 가정을 해보겠습니다. 사람은 차량의 라인업과 엠블럼을 보고 알지만 모델 입장에서 외형적인 특징을 뽑은 후 유사한 것을 분류해내는 것인데 시각의 기준에서 K5와 아반떼의 이미지적 특징이 같다면, 모델이 K5를 보고 아반떼라 부르는 것이 틀린건가요? 

다른 예로 2020년도 K5는 아반떼를 닮았지만, 2021년도 K5는 모하비를 닮았다고 합시다. 아예 다른 시각적 특징을 가진 20년식과 21년식 K5를 같은 라벨로 분류하는 것을 시각의 관점에서 정답이라 말할 수 있을까요?

4. 마지막으로 궁금한 점입니다! input img size가 현재 288*432로 고정입니다. 이로인해서 학습이 잘 안될 수도 있을까요? 이미지 resize를 해야할까요?

예를 들어 4032x2560 이미지에서 10x20 만큼의 영역에 차 이미지가 있는데, 이것을 288 x 432로 줄이면 차 이미지 영역은 가로 세로가 각각 288/4032 432/2560 비율로 감소할거고 대략 1/20 1/6로 줄어들게 될 것입니다. 이 경우 리사이징 된 입력 이미지에서 차 영역은 1x3 픽셀 정도일텐데 픽셀 3칸 가지고 분류를 해내는 것은 어렵기도 하지만 정답을 맞춰도 이상한 일이겠죠.

이러한 경우가 아니라면, 전체 이미지속 분류하고자 하는 대상 이미지가 차지하는 영역이 너무 작은게 아니라면 상관은 없습니다.

**Comment by JeongGod** [comment]
In GitLab by @jschoi on Nov 23, 2021, 01:07

Detectron2 성능이 안 나오는 것에 관련해서는 제가 VM에 코드를 올려두겠습니다!

**Comment by JeongGod** [comment]
In GitLab by @JeongGyu on Nov 23, 2021, 01:40

답변 감사드립니다! 내일 코치님 말씀대로 한 번 데이터를 수정해보도록 하겠습니다.
그리고 제가 jupyter를 열어서 학습시키는게 편해서 서버를 열어뒀습니다.
디코에 있는 도메인으로 접속하시고 비밀번호는 1q2w3e4r! 입니다! 
이렇게 늦은 시간에 답변해주셔서 다시 한 번 감사드립니다!:pray:
